File: javascript/0904-fruit-into-baskets.js
/**
 * @param {number[]} fruits
 * @return {number}
 */
var totalFruit = function (fruits) {
    let count = new Map();
    let [left, total, res] = [0, 0, 0];

    for (fruit of fruits) {
        count.set(fruit, (count.get(fruit) || 0) + 1);
        total++;

        while (count.size > 2) {
            let f = fruits[left];
            count.set(f, count.get(f) - 1);
            total -= 1;
            left += 1;
            if (!count.get(f)) {
                count.delete(f);
            }
        }
        res = Math.max(res, total);
    }

    return res;
};

File: go/0904-fruit-into-baskets.go
package main

func main() {

}

func totalFruit(fruits []int) int {
	count := map[int]int{}
	left, total, res := 0, 0, 0

	for _, fruit := range fruits {
		count[fruit]++
		total++

		for len(count) > 2 {
			f := fruits[left]
			count[f]--
			total--
			left++
			if count[f] == 0 {
				delete(count, f)
			}
		}
		res = max(res, total)
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}